config/yaml.go

```diff
@@ -26,32 +27,159 @@ import (
+		if err := t.validate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 // ListSlaves is the list of configurations of the slaves from the configuration
 // file.
-type ListSlaves map[string]*Slave
+type ListSlaves map[string]*Target
 
-// Slave defines the configuration parameters of a single slave.
+// Target defines the configuration parameters of a single slave.
 // Parity Values => N (None), E (Even), O (Odd)
 //
 // Default serial:
 // Baudrate: 19200, Databits: 8, Stopbits: 1, Parity: E
-type Slave struct {
-	Port          string   `yaml:"port"`
-	ID            byte     `yaml:"id"`
-	Timeout       int      `yaml:"timeout"`
-	Baudrate      int      `yaml:"baudrate"`
-	Databits      int      `yaml:"databits"`
-	Stopbits      int      `yaml:"stopbits"`
-	Parity        string   `yaml:"parity"`
-	KeepAlive     bool     `yaml:"keepAlive"`
-	DigitalInput  []string `yaml:"digitalIn"`
-	DigitalOutput []string `yaml:"digitalOut"`
-	AnalogInput   []string `yaml:"analogIn"`
-	AnalogOutput  []string `yaml:"analogOut"`
+type Target struct {
+	Name          string      `yaml:"name"`
+	Port          string      `yaml:"port"`
+	ID            byte        `yaml:"id"`
+	Timeout       int         `yaml:"timeout"`
+	Baudrate      int         `yaml:"baudrate"`
+	Databits      int         `yaml:"databits"`
+	Stopbits      int         `yaml:"stopbits"`
+	Parity        string      `yaml:"parity"`
+	KeepAlive     bool        `yaml:"keepAlive"`
+	DigitalInput  []MetricDef `yaml:"digitalIn"`
+	DigitalOutput []MetricDef `yaml:"digitalOut"`
+	AnalogInput   []MetricDef `yaml:"analogIn"`
+	AnalogOutput  []MetricDef `yaml:"analogOut"`
+}
+
+// RegisterAddr specifies the register in the possible output of _digital
+// output_, _digital input, _ananlog input, _analog output_.
+type RegisterAddr uint16
+
+// ModbusDataType is an Enum, representing the possible data types a register
+// value can be interpreted as.
+type ModbusDataType string
+
+func (t *ModbusDataType) validate() error {
+	if t == nil {
+		return fmt.Errorf("expected data type not to be nil")
+	}
+
+	for _, possibelType := range possibelModbusDataTypes {
+		if *t == possibelType {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("expected one of the following data types %v but got '%v'", possibelModbusDataTypes, *t)
+}
+
+// Parse parses the given byte slice based on the specified Modbus data type and
+// returns the parsed value as a float64 (Prometheus exposition format).
+//
+// TODO: Does this belong here?
+// TODO: Handle Endianness.
+func (d *MetricDef) Parse(rawData [2]byte) (float64, error) {
+	if d == nil {
+		return 0, fmt.Errorf("expected metric definition not to be nil")
+	}
+
+	switch d.DataType {
+	case ModbusFloat16:
+		panic("implement")
+	case ModbusInt16:
+		{
+			i := binary.BigEndian.Uint16(rawData[:])
+			return float64(int16(i)), nil
+		}
+	case ModbusUInt16:
+		{
+			i := binary.BigEndian.Uint16(rawData[:])
+			return float64(i), nil
+		}
+	case ModbusBool:
```

Comment:
@RichiH I have added a prove-of-concept implementation of the _boolean_ data type. Users specify a _register_ and a _bit position_ within the register.

Multiple questions arise on how to address bit positions within a register (16 bit):

-  Should the _bit position_ be zero indexed?

- Should the _bit position_ start with the least or most significant bit?

- What endianness do we use here?

Can you post a couple of examples with _16 bit input_, _bit position_ and _expected output_ here?

Reply:
I think the only valid interpretation of these 16 bits is as a bit vector, which also sidesteps/answers these nicely. It's

* zero indexed (think of it as an offset to the base address)
* network order (first bit on the wire is 0)
* network order implies big endian, but this is not "valid", or meaningful, for bit vectors anyway.

This approach also matches most closely with what I saw in the field.

Given the above, do you still need samples?